runtime/lua/vim/lsp/buf.lua

```diff
@@ -192,17 +192,25 @@ end
 ---     - range (table|nil) Range to format.
 ---         Table must contain `start` and `end` keys with {row,col} tuples using
 ---         (1,0) indexing.
+---         Can also be a list of tables that contain `start` and `end` keys as described above,
+---         in which case `textDocument/rangesFormatting` support is required.
 ---         Defaults to current selection in visual mode
----         Defaults to `nil` in other modes, formatting the full buffer
+---         and to `nil` in other modes, formatting the full buffer
 function M.format(options)
   options = options or {}
   local bufnr = options.bufnr or api.nvim_get_current_buf()
   local mode = api.nvim_get_mode().mode
   local range = options.range
+  -- Try to use visual selection if no range is given
   if not range and mode == 'v' or mode == 'V' then
     range = range_from_selection(bufnr, mode)
   end
-  local method = range and ms.textDocument_rangeFormatting or ms.textDocument_formatting
+  -- List of ranges => textDocument/rangesFormatting
+  -- Single range   => textDocument/rangeFormatting
+  -- No range       => textDocument/formatting
+  local passed_multiple_ranges = (range and #range ~= 0 and type(range[1]) == 'table')
+  local method = (passed_multiple_ranges and ms.textDocument_rangesFormatting)
+    or (range and ms.textDocument_rangeFormatting or ms.textDocument_formatting)
```

Comment:
I find this too complex to read. Let's rewrite using plain if-else?

Reply:
Also can we have capability check? (either explicitly or implicitly) Though not mandatory, it'd be even better if this can be checked by tests.

Reply:
> I find this too complex to read. Let's rewrite using plain if-else?
Done!

> Also can we have capability check? (either explicitly or implicitly) Though not mandatory, it'd be even better if this can be checked by tests.

Our call to `vim.lsp.get_clients` already filters by method, so there's already a (somewhat hidden) capability check there,